numba/dppl/dppl_lowerer.py

```diff
@@ -962,16 +962,166 @@ def __init__(self, *args, **kwargs):
+
+    from numba.core.dispatcher import _DispatcherBase
+    from numba.core.types.functions import Function, Dispatcher
+    from numba.core.bytecode import FunctionIdentity
+    from numba.core.typing.templates import Signature
+    from numba.dppl.compiler import DPPLFunctionTemplate
+    from numba.core.compiler import CompileResult
+    from numba.np.ufunc.dufunc import DUFunc
+    from ctypes import _CFuncPtr
+    from cffi.api import FFI
+    from types import ModuleType
+    from numba.core.types.abstract import Type
+
+    if isinstance(obj, (FunctionIdentity, _DispatcherBase, Function, Type, Dispatcher, ModuleType,
+                        Signature, DPPLFunctionTemplate, CompileResult,
+                        DUFunc, _CFuncPtr, FFI,
+                        type, str, bool, type(None))):
+        return obj
+
+    from numba.core.ir import Global, FreeVar
+    from numba.core.ir import FunctionIR
+    from numba.core.postproc import PostProcessor
+    from numba.core.funcdesc import FunctionDescriptor
+
+    if isinstance(obj, FunctionDescriptor):
+        cpy = FunctionDescriptor(native=obj.native, modname=obj.modname, qualname=obj.qualname,
+                                 unique_name=obj.unique_name, doc=obj.doc,
+                                 typemap=relatively_deep_copy(obj.typemap, memo),
+                                 restype=obj.restype,
+                                 calltypes=relatively_deep_copy(obj.calltypes, memo),
+                                 args=obj.args, kws=obj.kws, mangler=None,
+                                 argtypes=relatively_deep_copy(obj.argtypes, memo),
+                                 inline=obj.inline, noalias=obj.noalias, env_name=obj.env_name,
+                                 global_dict=obj.global_dict)
+        cpy.mangled_name = obj.mangled_name
+
+        memo[obj_id] = cpy
+
+        return cpy
+
+    if isinstance(obj, FunctionIR):
+        pp = PostProcessor(obj)
+        pp.run()
+        cpy = FunctionIR(blocks=relatively_deep_copy(obj.blocks, memo),
+                         is_generator=relatively_deep_copy(obj.is_generator, memo),
+                         func_id=relatively_deep_copy(obj.func_id, memo),
+                         loc=obj.loc,
+                         definitions=relatively_deep_copy(obj._definitions, memo),
+                         arg_count=obj.arg_count,
+                         arg_names=relatively_deep_copy(obj.arg_names, memo))
+        pp = PostProcessor(cpy)
+        pp.run()
+
+        memo[obj_id] = cpy
+
+        return cpy
+
+    if isinstance(obj, Global):
+        cpy = Global(name=obj.name, value=obj.value, loc=obj.loc)
+        memo[obj_id] = cpy
+
+        return cpy
+
+    if isinstance(obj, FreeVar):
+        cpy = FreeVar(index=obj.index, name=obj.name, value=obj.value, loc=obj.loc)
+        memo[obj_id] = cpy
+
+        return cpy
+
+    if isinstance(obj, list):
+        cpy = copy.copy(obj)
+        cpy.clear()
```

Comment:
Why do you copy and immediately clear?

Reply:
subclasses can have additional members which we need to be copied. At the same moment `copy` doesn't apply `relatively_deep_copy` to list items